src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -1907,13 +1913,87 @@ public static bool TryGetLibraryExport(IntPtr handle, string name, out IntPtr ad
             return address != IntPtr.Zero;
         }
 
+        /// <summary>
+        /// Map from assembly to native-library-resolution-callback.
+        /// Generally interop specific fields and properties are not added to assembly.
+        /// Therefore, this table uses weak assembly pointers to indirectly achieve 
+        /// similar behavior.
+        /// </summary>
+        public static ConditionalWeakTable<Assembly, Func<string, Assembly, DllImportSearchPath?, IntPtr>> nativeDllResolveMap;
+
+        /// <summary>
+        /// Register a callback for resolving native library imports from an assembly
+        /// This per-assembly callback is the first attempt to resolve native library loads 
+        /// initiated by this assembly.
+        ///
+        /// Only one callback can be registered per assembly. Trying to register a second
+        /// callback fails (with the return value false).
+        ///
+        /// The callback method itself takes the following parameters
+        ///   - The name of the library to be loaded (string)
+        ///   - The assembly initiating the native library load
+        ///   - The DllImportSearchPath flags from the attributes on the PInvoke, if any (null otherwise).
+        ///     This argument doesn't include the flags from any attributes on the assembly itself.
+        /// and returns
+        ///   - The handle to the loaded native library (on success) or null (on failure)
+        /// The parameters on this callback are such that they can be directly passed to 
+        /// Marhall.LoadLibrary(libraryName, assembly, dllImportSearchPath) to approximately achieve 
+        /// the default load behavior.
+        ///
+        /// </summary>
+        /// <param name="assembly">The assembly for which the callback is registered</param>
+        /// <param name="callBack">The callback to register</param>
+        /// <exception cref="System.ArgumentNullException">If assembly or callback is null</exception>
+        /// <returns>True on success, false otherwise</returns>  
+        public static bool RegisterDllImportResolver(Assembly assembly, Func<string, Assembly, DllImportSearchPath?, IntPtr> callBack)
+        {
+            if (assembly == null)
+                throw new ArgumentNullException(nameof(assembly));
+            if (callBack == null)
+                throw new ArgumentNullException(nameof(callBack));
+            if (!(assembly is RuntimeAssembly))
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+
+            Func<string, Assembly, DllImportSearchPath?, IntPtr> existingCallback;
+            if (nativeDllResolveMap.TryGetValue(assembly, out existingCallback))
+            {
+                return false;
+            }
+
+            nativeDllResolveMap.Add(assembly, callBack);
```

Comment:
This has race condition: if the method is called from multiple threads on same assembly, we may end up throwing here.